File: sdk/python/runtime_client.py
"""
Python SDK middleware for REM Verified Inference Runtime.

Features (MVP):
- Non-streaming and streaming chat completions via OpenAI-compatible proxy
- Header management (x-api-key, x-byo-key, x-purpose, x-request-id)
- Basic replay artifacts: replay_stub.json lines and replay.sh script

Usage:
  from mvp.sdk.python.runtime_client import RuntimeClient
  rc = RuntimeClient(base_url="http://localhost:8080", api_key=..., byo_key=..., purpose="regulatory_filing")
  # non-streaming
  res = anyio.run(rc.chat_completions, {"model":"gpt-4o-mini","messages":[{"role":"user","content":"hi"}]})
  # streaming
  async def on_evt(evt):
      print(evt["event"], evt.get("data"))
  await rc.chat_completions({"model":"gpt-4o-mini","messages":[...] , "stream": True}, on_event=on_evt)
"""
from __future__ import annotations

import os
import json
import time
import uuid
from typing import Any, Dict, Optional, Callable, Awaitable

import httpx


class RuntimeClient:
    def __init__(self, base_url: str, api_key: Optional[str] = None, byo_key: Optional[str] = None, purpose: Optional[str] = None) -> None:
        if not base_url:
            raise ValueError("base_url required")
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key or os.getenv("API_KEY")
        self.byo_key = byo_key or os.getenv("BYO_KEY")
        self.purpose = purpose or os.getenv("RUNTIME_PURPOSE")

    def _headers(self, req_id: Optional[str] = None) -> Dict[str, str]:
        h = {"Content-Type": "application/json"}
        if req_id:
            h["x-request-id"] = req_id
        if self.api_key:
            h["x-api-key"] = self.api_key
        if self.byo_key:
            h["x-byo-key"] = self.byo_key
        if self.purpose:
            h["x-purpose"] = self.purpose
        return h

    async def chat_completions(self, payload: Dict[str, Any], on_event: Optional[Callable[[Dict[str, Any]], Awaitable[None]]] = None) -> Optional[Dict[str, Any]]:
        """If payload.stream is True, invoke streaming and call on_event for SSE frames; otherwise return JSON.
        on_event receives dict: {event, data, raw}
        """
        stream = bool(payload.get("stream"))
        rid = str(uuid.uuid4())
        url = f"{self.base_url}/proxy/openai/v1/chat/completions"
        headers = self._headers(rid)
        if not stream:
            async with httpx.AsyncClient(timeout=60) as hc:
                resp = await hc.post(url, headers=headers, json=payload)
                resp.raise_for_status()
                data = resp.json()
            self._write_replay_stub(rid, payload, headers, streamed=False)
            return data

        # streaming path
        async with httpx.AsyncClient(timeout=None) as hc:
            async with hc.stream("POST", url, headers=headers, json=payload) as resp:
                resp.raise_for_status()
                chunk_hashes = []
                async for event in _iter_sse(resp):
                    # collect chunk hashes for manifest/chunk
                    if event.get("event") == "rem.chunk":
                        try:
                            chunk_hashes.append(event.get("data", {}).get("sha256"))
                        except Exception:
                            pass
                    if on_event:
                        await on_event(event)
        self._write_replay_stub(rid, payload, headers, streamed=True, chunk_hashes=[h for h in chunk_hashes if h])
        return None

    def _write_replay_stub(self, request_id: str, payload: Dict[str, Any], headers: Dict[str, str], streamed: bool, chunk_hashes: Optional[list[str]] = None) -> None:
        record = {
            "ts": time.time(),
            "request_id": request_id,
            "endpoint": "/proxy/openai/v1/chat/completions",
            "payload": {k: payload.get(k) for k in ("model", "messages", "temperature", "seed", "tools", "stream")},
            "headers": {k: headers.get(k) for k in ("x-request-id", "x-purpose") if headers.get(k)},
            "streamed": streamed,
            "chunk_hashes": chunk_hashes or [],
        }
        try:
            os.makedirs("audit/repro", exist_ok=True)
            # append line
            with open("audit/repro/replay_stub.jsonl", "a", encoding="utf-8") as f:
                f.write(json.dumps(record, ensure_ascii=False) + "\n")
            # ensure replay.sh exists
            sh = (
                "#!/usr/bin/env bash\n"
                "# Minimal replay helper (placeholder)\n"
                "echo 'Provide an audit JSONL line to audit/repro/inference_replay.py'\n"
            )
            p = "audit/repro/replay.sh"
            if not os.path.exists(p):
                with open(p, "w", encoding="utf-8") as f:
                    f.write(sh)
        except Exception:
            return


async def _iter_sse(resp: httpx.Response):
    """Parse a server-sent-events stream from httpx response.
    Yields dicts: {event, data, raw}
    """
    event = "message"
    data_lines = []
    async for line in resp.aiter_lines():
        if line is None:
            continue
        raw = line
        s = line.rstrip("\r")
        if s == "":
            if data_lines:
                raw_data = "\n".join(data_lines)
                try:
                    parsed = json.loads(raw_data)
                except Exception:
                    parsed = raw_data
                yield {"event": event or "message", "data": parsed, "raw": raw_data}
            event = "message"
            data_lines = []
            continue
        if s.startswith(":"):
            continue
        if s.startswith("event:"):
            event = s[6:].strip()
            continue
        if s.startswith("data:"):
            data_lines.append(s[5:].lstrip())
            continue
        # ignore other fields

File: sdk/js/runtimeClient.js
// Minimal JS SDK for REM Verified Inference Runtime
// Node >=18 (global fetch) or browsers

export class RuntimeClient {
  constructor({ baseUrl, apiKey, byoKey } = {}) {
    if (!baseUrl) throw new Error("baseUrl required");
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey || process.env.API_KEY || undefined;
    this.byoKey = byoKey || process.env.BYO_KEY || undefined;
  }

  _headers(extra = {}) {
    const h = { "Content-Type": "application/json", ...extra };
    if (this.apiKey) h["x-api-key"] = this.apiKey;
    if (this.byoKey) h["x-byo-key"] = this.byoKey; // upstream key
    return h;
  }

  // Non-streaming chat completions (OpenAI-compatible payload)
  async chatCompletions(payload) {
    const url = `${this.baseUrl}/proxy/openai/v1/chat/completions`;
    const resp = await fetch(url, {
      method: "POST",
      headers: this._headers(),
      body: JSON.stringify({ ...payload, stream: false }),
    });
    if (!resp.ok) {
      const msg = await safeText(resp);
      throw new Error(`Proxy error ${resp.status}: ${msg}`);
    }
    return resp.json();
  }

  // Streaming chat completions (SSE). onEvent receives { event, data, raw }
  async chatCompletionsStream(payload, onEvent) {
    const url = `${this.baseUrl}/proxy/openai/v1/chat/completions`;
    const resp = await fetch(url, {
      method: "POST",
      headers: this._headers(),
      body: JSON.stringify({ ...payload, stream: true }),
    });
    if (!resp.ok || !resp.body) {
      const msg = await safeText(resp);
      throw new Error(`Proxy error ${resp.status}: ${msg}`);
    }
    await parseEventStream(resp.body, onEvent);
  }
}

async function safeText(resp) {
  try {
    return await resp.text();
  } catch {
    return "";
  }
}

// Minimal SSE parser for ReadableStream
async function parseEventStream(stream, onEvent) {
  const reader = stream.getReader();
  const decoder = new TextDecoder();
  let buf = "";
  // Basic SSE framing: lines with optional 'event:' and 'data:'; events separated by blank line
  let cur = { event: "message", data: [] };
  while (true) {
    const { value, done } = await reader.read();
    if (done) break;
    buf += decoder.decode(value, { stream: true });
    let idx;
    while ((idx = buf.indexOf("\n")) >= 0) {
      const line = buf.slice(0, idx);
      buf = buf.slice(idx + 1);
      const s = line.replace(/\r$/, "");
      if (s === "") {
        // dispatch
        if (cur.data.length > 0) {
          const raw = cur.data.join("\n");
          let parsed = raw;
          try { parsed = JSON.parse(raw); } catch {}
          onEvent && onEvent({ event: cur.event || "message", data: parsed, raw });
        }
        cur = { event: "message", data: [] };
        continue;
      }
      if (s.startsWith(":")) continue; // comment
      if (s.startsWith("event:")) {
        cur.event = s.slice(6).trim();
        continue;
      }
      if (s.startsWith("data:")) {
        cur.data.push(s.slice(5).trimStart());
        continue;
      }
      // ignore other fields for MVP
    }
  }
}

export default RuntimeClient;


